### ISSUE TITLE
Some problems in experiments

### ISSUE THREAD

**Issue by woodenchild95** [optsuite/LOZO#1]
Dear authors, I recently came across your LOZO work, and I think it is an outstanding piece of work. Both in terms of implementation and its efficiency, it has remarkable properties. While attempting to reproduce its experimental results, I encountered an issue and would like to inquire about the possible causes:

In my testing experiment with OPT-1.3B-sst2, I set v=1, meaning each sample in a batch has independent U and V sampling. The effect of v=1 has already been very good. However, I noticed that in the loss curve reported on the right side of Figure 4 on page 21 of the paper, v=1 almost does not converge. I would like to ask about the possible reasons for this and what the corresponding hyperparameters are. In my reproduction, I directly adopted the hyperparameter selection from mezo: lr=1e-6, perturbation e=1e-3, and rank=1, which achieves almost the same effect as mezo. Additionally, when rank=4, the performance even becomes worse. I would like to ask if you have encountered this phenomenon in your experiments. It seems related to the variance of the subspace projection onto the true gradient, but unfortunately, I did not observe any reduction in variance.

**Comment by LiyuanCao** [comment]
We appreciate your interest in our paper.

We would like to clarify a few points regarding the experiment depicted in the right panel of Figure 4 on page 21. The results shown there are from an ablation study conducted with rank = 2 and different values of 𝜈. Other hyperparameters remain consistent with those listed in Table 5. We apologize for any confusion this may have caused.

Regarding your suggestion that setting 𝜈 = 1 with rank = 1 should yield performance comparable to MeZO, we reran this experiment using the settings you specified. However, the results did not achieve the same level of performance as MeZO. Could you please provide a more detailed log of your experiments? This information would help us investigate the issue more thoroughly. In our experiments, we observed that setting 𝜈 too low often leads to insufficient subspace optimization, which generally results in suboptimal performance.

Additionally, you noted that increasing the rank leads to a drop in performance. This observation is consistent with our findings. As shown in Table 6 on page 22, our algorithm attains the best results with a rank of 2, while the performance declines when the rank is increased to 8.

Regarding this phenomenon, our explanation is that a higher rank does not necessarily result in reduced variance in the gradient estimator. Although a larger rank may lower the variance of the subspace projection onto the true gradient, it simultaneously increases the complexity of the inner optimization problem (as described in Eq. (14a) on page 6) due to its expanded scale. This increased complexity can counteract the expected variance reduction, thus not necessarily improving overall performance. Therefore, we believe that using a relatively smaller rank is optimal, as it both preserves the low-rank structure of the gradient estimator and ensures minimal memory overhead. If you wish to improve performance with a higher rank, we recommend experimenting with a larger value of 𝜈 and a reduced learning rate to better handle the increased complexity of the inner problem.

If you have additional questions, please feel free to open an issue on our GitHub repository. We are happy to assist further. Thank you again for your engagement!

**Comment by woodenchild95** [comment]
@LiyuanCao Thank you for your prompt response. Regarding the experiment with v=1, in fact, each batch re-samples U and V anew. Here’s my test implementation:

‘’‘
def zo_perturb_parameters(self, random_seed=None, scaling_factor=1):
        """
        Perturb the parameters with random vector z.
        Input: 
        - random_seed: random seed for MeZO in-place perturbation (if it's None, we will use self.zo_random_seed)
        - scaling_factor: theta = theta + scaling_factor * z * eps
        """

        # Set the random seed to ensure that we sample the same z for perturbation/update
        torch.manual_seed(random_seed if random_seed is not None else self.zo_random_seed)
        
        for name, param in self.named_parameters_to_optim:
            ### Mezo
            # z = torch.normal(mean=0, std=1, size=param.data.size(), device=param.data.device, dtype=param.data.dtype)
            
            ### LOZO
            if param.data.dim() == 2:
                ml, nl = param.data.size()
                u_size = torch.Size([ml, self.low_rank])
                v_size = torch.Size([self.low_rank, nl])
                ul = torch.normal(mean=0, std=1, size=u_size, device=param.data.device, dtype=param.data.dtype)
                vl = torch.normal(mean=0, std=1, size=v_size, device=param.data.device, dtype=param.data.dtype)
                z = torch.mm(ul, vl)
            elif param.data.dim() == 1:
                z = torch.normal(mean=0, std=1, size=param.data.size(), device=param.data.device, dtype=param.data.dtype)
            else:
                raise NotImplementedError("Dimension is larger than 2! Please consider high demension tensors.")

            param.data = param.data + scaling_factor * z * self.args.zo_eps / self.scalar
’‘’
Below is the segment related to LOZO extracted from my modifications to the MeZO code.

My test environment is:
Pytorch 2.0.1, Cuda 11.7, Driver Version 560.94, Transformers 4.25.1
My test setups:
lr = 1e-6, eps = 1e-3, and I use float16 type. 

I am trying to further accelerate LOZO's convergence, but Im not sure that my implementation strictly follows your algorithm design. This setup seems to yield results similar to MeZO. The log file from MeZO doesn’t seem to have much additional information; the outputs mostly show parameters related to the OPT model. I used MeZO’s default settings, with modifications limited to the code parts mentioned above.

**Comment by itongggg** [comment]
@woodenchild95  thank you for your further clarification,
 could you provide your gradient update part code.

**Comment by woodenchild95** [comment]
@itongggg Thank you very much for your willingness to review my related code. I think that the implementation aligns with your algorithm's design, but I am unsure if some details might have been overlooked. Below, I have included the modifications I made to the trainer.py file in the Mezo codebase.

The zo_perturb_parameters function:

      def zo_perturb_parameters(self, random_seed=None, scaling_factor=1):
              # Set the random seed to ensure that we sample the same z for perturbation/update
              torch.manual_seed(random_seed if random_seed is not None else self.zo_random_seed)
              
              for name, param in self.named_parameters_to_optim:
                  ### Lozo
                  if param.data.dim() == 2:
                      ml, nl = param.data.size()
                      u_size = torch.Size([ml, self.low_rank])
                      v_size = torch.Size([self.low_rank, nl])
                      ul = torch.normal(mean=0, std=1, size=u_size, device=param.data.device, dtype=param.data.dtype)
                      vl = torch.normal(mean=0, std=1, size=v_size, device=param.data.device, dtype=param.data.dtype)
                      z = torch.mm(ul, vl)
                  elif param.data.dim() == 1:
                      z = torch.normal(mean=0, std=1, size=param.data.size(), device=param.data.device, dtype=param.data.dtype)
                  else:
                      raise NotImplementedError("Dimension is larger than 2! Please consider high demension tensors.")
      
                  param.data = param.data + scaling_factor * z * self.args.zo_eps / self.scalar

And the zo_update function:
        
    def zo_update(self, model):
                args = self.args
        
                # Reset the random seed for sampling zs
                torch.manual_seed(self.zo_random_seed)     
        
                for name, param in self.named_parameters_to_optim:
                    # Resample z
                    ### Lozo
                    if param.data.dim() == 2:
                        ml, nl = param.data.size()
                        u_size = torch.Size([ml, self.low_rank])
                        v_size = torch.Size([self.low_rank, nl])
                        ul = torch.normal(mean=0, std=1, size=u_size, device=param.data.device, dtype=param.data.dtype)
                        vl = torch.normal(mean=0, std=1, size=v_size, device=param.data.device, dtype=param.data.dtype)
                        z = torch.mm(ul, vl)
                    elif param.data.dim() == 1:
                        z = torch.normal(mean=0, std=1, size=param.data.size(), device=param.data.device, dtype=param.data.dtype)
                    else:
                        z = None
                        raise NotImplementedError("Dimension is larger than 2! Please consider high demension tensors.")
                    
                    if "bias" not in name and "layer_norm" not in name and "layernorm" not in name:
                        param.data = param.data - self._get_learning_rate() * (self.projected_grad * z / self.scalar + args.weight_decay * param.data)
                    else:
                        param.data = param.data - self._get_learning_rate() * (self.projected_grad * z / self.scalar)
        
                self.lr_scheduler.step()


My current testing of LOZO is based on the code above, and the performance at rank=1 is excellent, comparable to MeZO and even surpassing it in certain results. Additionally, I have also tried bypassing the computation of $u$ and $v$ by directly incorporating them into the update steps. This approach eliminates the need to store the intermediate value $z$, instead computing it through broadcasting, which is a memory-efficient operation. Despite this adjustment, it still achieves excellent training performance. The scalar term is selected as $\sqrt{r}$, which ensures that the variance of the entire gradient is consistent with the method used by MeZO. However, when $r=1$, the scalar becomes $1$, so using $\sqrt{r}$ and $r$ (as in your paper's form) is equivalent.

**Comment by itongggg** [comment]
@woodenchild95  

Thank you for your detailed description. I may have misunderstood some aspects. In our paper, we found that setting `rank=1` can indeed be competitive compared to MeZO under certain datasets. However, setting `step_interval` $\nu$ to 1 does not exhibit the same performance.

Do you mean that setting both `rank=1` and $\nu=1$ still exhibits comparable performance? Additionally, I want to clarify that our ablation study does not include the setting `rank=1` and $\nu=1$. Instead, we included the setting `rank=2` and $\nu=1$. Perhaps increasing the rank can undermine the performance.

I also ran experiments with the setting `rank=1` and `$\nu=1$` on the SST-2 dataset using opt-1.3b, achieving an accuracy of 92.9%, which aligns with your results. However, as our paper suggests, this setting $\nu=1$ may undermine performance when using a higher rank.

**Comment by woodenchild95** [comment]
@itongggg hi, rank=2 and v=1 still works for me on the above codes and its performance is still good. So, I still couldn’t successfully reproduce the curve for $v=1$ in Figure 4 (right) on page 21 of the paper. Could this be an issue with the learning rate selection? If there are no fundamental errors in my implementation, I currently think it might be related to the choice of learning rate or scalar parameters ($1/r$ or $1/\sqrt{r}$). Thank you very much for your response—I now have a clearer understanding of the algorithm’s performance.

**Comment by itongggg** [comment]
@woodenchild95  

To simplify the code implementation, we did not explicitly scale the learning rate in our `zo_update` function. As discussed in our paper, this scaling is inherently reflected in the learning rate itself. For example, when we set the learning rate to `1e-7` with `rank=2`, it effectively operates as `2e-7`. For Figure 4 (right) on page 21 of the paper, we consistently use a learning rate of `1e-7`. 
We hope this clarification helps! Thank you again for your thoughtful feedback and for engaging with our work.
